File: prolog/triangle/triangle.pl
triangle(Side1, Side2, Side3, "equilateral") :- 
	is_triangle(Side1, Side2, Side3),
	equilateral(Side1, Side2, Side3), !.
triangle(Side1, Side2, Side3, "isosceles") :- 
	is_triangle(Side1, Side2, Side3),
	isosceles(Side1, Side2, Side3), !.
triangle(Side1, Side2, Side3, "scalene") :- 
	is_triangle(Side1, Side2, Side3),
	scalene(Side1, Side2, Side3), !.

is_triangle(Side1,Side2,Side3) :- 
	no_zero_sides(Side1,Side2,Side3),
	equality(Side1,Side2,Side3).

no_zero_sides(Side1,Side2,Side3) :-
	Side1 > 0,
	Side2 > 0,
	Side3 > 0.

equality(Side1,Side2,Side3) :-
	Side1 + Side2 >= Side3,
	Side1 + Side3 >= Side2, 
	Side2 + Side3 >= Side1.

equilateral(X,X,X).

isosceles(X,X,_).
isosceles(X,_,X).
isosceles(_,X,X).

scalene(X,Y,Z) :-
	X =\= Y,
	X =\= Z,
	Y =\= Z.

File: kotlin/triangle/src/main/kotlin/Triangle.kt
class Triangle<out T : Number>(val a: T, val b: T, val c: T) {
    fun no_zero_sides(a: Double, b: Double, c: Double): Boolean {
        return a > 0.0 && b > 0.0 && c > 0.0
    }

    fun equality(a: Double, b: Double, c: Double): Boolean {
        return a + b >= c
                && a + c >= b
                && b + c >= a
    }

    fun is_triangle(a: Double, b: Double, c: Double): Boolean {
        return no_zero_sides(a, b, c) && equality(a, b, c)
    }

    init {
        require(is_triangle(a.toDouble(), b.toDouble(), c.toDouble())) { "Not a triangle" }
    }
    
    val isEquilateral: Boolean = a == b && b == c
    val isIsosceles: Boolean = ((a == b || a == c) && b != c) || (b == c && a != b) || isEquilateral
    val isScalene: Boolean = a != b && a != c && b != c
}
